src/feasibility_checker.jl

```diff
@@ -0,0 +1,161 @@
+    violated_constraints = Dict{Any,Float64}()
+    for (F, S) in list_of_constraint_types(model)
+        # This loop is not type-stable because `F` and `S` change, but it should
+        # be fine because no one should be using this in performance-critical
+        # code.
+        for con in all_constraints(model, F, S)
+            obj = constraint_object(con)
+            d = _distance_to_set(value.(obj.func, point_f), obj.set)
+            if d > atol
+                violated_constraints[con] = d
+            end
+        end
+    end
+    if num_nl_constraints(model) > 0
+        evaluator = NLPEvaluator(model)
+        MOI.initialize(evaluator, Symbol[])
+        g = zeros(num_nl_constraints(model))
+        MOI.eval_constraint(evaluator, g, point_f.(all_variables(model)))
+        for (i, con) in enumerate(model.nlp_data.nlconstr)
+            d = max(0.0, con.lb - g[i], g[i] - con.ub)
+            if d > atol
+                cref = ConstraintRef(
+                   model,
+                   NonlinearConstraintIndex(i),
+                    ScalarShape(),
+                )
+                violated_constraints[cref] = d
+            end
+        end
+    end
+    return length(violated_constraints) == 0 ? nothing : violated_constraints
+end
+
+function _distance_to_set(::Any, set::MOI.AbstractSet)
+    error(
+        "Feasibility checker for set type $(typeof(set)) has not been " *
+        "implemented yet."
+    )
+end
+
+###
+### MOI.AbstractScalarSets
+###
+
+function _distance_to_set(x::T, set::MOI.LessThan{T}) where {T<:Real}
+    return max(x - set.upper, zero(T))
+end
+
+function _distance_to_set(x::T, set::MOI.GreaterThan{T}) where {T<:Real}
+    return max(set.lower - x, zero(T))
+end
+
+function _distance_to_set(x::T, set::MOI.EqualTo{T}) where {T<:Number}
+    return abs(set.value - x)
+end
+
+function _distance_to_set(x::T, set::MOI.Interval{T}) where {T<:Real}
+    return max(x - set.upper, set.lower - x, zero(T))
+end
+
+function _distance_to_set(x::T, ::MOI.ZeroOne) where {T<:Real}
+    return min(abs(x - zero(T)), abs(x - one(T)))
+end
+
+function _distance_to_set(x::T, ::MOI.Integer) where {T<:Real}
+    return abs(x - round(x))
+end
+
+function _distance_to_set(x::T, set::MOI.Semicontinuous{T}) where {T<:Real}
+    return min(max(x - set.upper, set.lower - x, zero(T)), abs(x))
+end
+
+function _distance_to_set(x::T, set::MOI.Semiinteger{T}) where {T<:Real}
+    d = max(
+        ceil(set.lower) - x,
+        x - floor(set.upper),
+        abs(x - round(x)),
+    )
+    return min(d, abs(x))
+end
+
+###
+### MOI.AbstractVectorSets
+###
+
+function _check_dimension(v::AbstractVector, s)
+    if length(v) != MOI.dimension(s)
+        throw(DimensionMismatch("Mismatch between value and set"))
+    end
+    return
+end
+
+function _distance_to_set(
+    x::Vector{T},
+    set::MOI.Nonnegatives,
+) where {T<:Real}
+    _check_dimension(x, set)
+    return LinearAlgebra.norm(max(-xi, zero(T)) for xi in x)
```

Comment:
If we go with an non-conic feasibility checker, it would make work sense to use infinity norm here to be consistent with the scalarized constraints. Otherwise, you would get different results whether you use functionize/scalarize bridges or not.

Reply:
> you would get different results whether you use functionize/scalarize bridges or not.

Yes. But there are going to be weird artifacts due to the tolerance no matter how we pick the distance. I'd rather have a single definition "The Euclidean distance between the point and the set," rather than defining different distances for different sets.

Reply:
It could also be the infinite norm for any set, no?

Reply:
Yes. But it could also be the L2 norm. As a standard definition, L2 makes the most sense for most sets, except where it is difficult to compute (e.g., PSDCone). The distance depends on the formulation, but that shouldn't be surprising. The formulations are different. The tolerances are also absolute, so they're not scale-invariant.

Reply:
L2 is not that difficult for PSD. There just isn't a closed formula for that. Moreover, L2 is probably the easiest for PSD.

On the main discussion... I think people will end up asking for alternate distances, but that can be done in the sequence.